pytorch_lightning/plugins/io/torch_plugin.py

```diff
@@ -31,7 +33,10 @@ class TorchCheckpointIO(CheckpointIO):
 
     def save_checkpoint(self, checkpoint: Dict[str, Any], path: _PATH, storage_options: Optional[Any] = None) -> None:
         fs = get_filesystem(path)
-        fs.makedirs(os.path.dirname(path), exist_ok=True)
+        dirname = os.path.dirname(path)
+        if fs.isdir(dirname) and len(fs.ls(dirname)) > 0:
+            warning_cache.warn(f"Checkpoint directory {dirname} exists and is not empty.")
+        fs.makedirs(dirname, exist_ok=True)
```

Comment:
i don't prefer this check here as much becaus if we're training with a plugin that supports saving sharded state, we still run the isdir check on all ranks. this will run into the network issues we saw earlier

@jjenniferdai what do you think about moving this check into `on_pretrain_routine_start` in the trainer, after `self.__resolve_ckpt_dir` finishes?

Reply:
good point I wasn't sure about trading off point 1 with fs code io plugin consolidation, but for point 2 this shouldn't live here.

what about here https://github.com/PyTorchLightning/pytorch-lightning/pull/9615#discussion_r713353601 - at the end of `__resolve_ckpt_dir`, within existing rank if case and before makedirs

Reply:
there's potential duplication because we need to handle the early return here: https://github.com/PyTorchLightning/pytorch-lightning/blob/8fcdcb598bbf0fbe6cda92415d1d69eb1e2d0aa0/pytorch_lightning/callbacks/model_checkpoint.py#L595-L598

which is why i prefer this sequence:
```
__resolve_ckpt_dir -> only initializes attributes on the checkpoint callback. does no IO
__check_if_dir_not_empty -> run this only on global rank 0 for the warning
```
and leave the `mkdirs` creation to the IO plugin

what do you think?

Reply:
op forgot about the early return, yep this makes more sense!